exporter/splunkhecexporter/client.go

```diff
@@ -178,74 +196,115 @@ func (c *client) pushLogDataInBatches(ctx context.Context, ld pdata.Logs, send f
 		ills := rls.At(i).InstrumentationLibraryLogs()
 		for j := 0; j < ills.Len(); j++ {
 			logs := ills.At(j).Logs()
-			for k := 0; k < logs.Len(); k++ {
-				if bufFront == nil {
-					bufFront = &logIndex{resource: i, library: j, record: k}
-				}
-
-				// Parsing log record to Splunk event.
-				event := mapLogRecordToSplunkEvent(res, logs.At(k), c.config, c.logger)
-				// JSON encoding event and writing to buffer.
-				if err := encoder.Encode(event); err != nil {
-					permanentErrors = append(permanentErrors, consumererror.Permanent(fmt.Errorf("dropped log event: %v, error: %v", event, err)))
-					continue
-				}
-
-				// Continue adding events to buffer up to capacity.
-				// 0 capacity is interpreted as unknown/unbound consistent with ContentLength in http.Request.
-				if buf.Len() <= int(bufCap) || bufCap == 0 {
-					// Tracking length of event bytes below capacity in buffer.
-					bufLen = buf.Len()
-					continue
-				}
-
-				tmpBuf.Reset()
-				// Storing event bytes over capacity in buffer before truncating.
-				if bufCap > 0 {
-					if over := buf.Len() - bufLen; over <= int(bufCap) {
-						tmpBuf.Write(buf.Bytes()[bufLen:buf.Len()])
-					} else {
-						permanentErrors = append(permanentErrors, consumererror.Permanent(
-							fmt.Errorf("dropped log event: %s, error: event size %d bytes larger than configured max content length %d bytes", string(buf.Bytes()[bufLen:buf.Len()]), over, bufCap)))
-					}
-				}
-
-				// Truncating buffer at tracked length below capacity and sending.
-				buf.Truncate(bufLen)
-				if buf.Len() > 0 {
-					if err := send(ctx, buf); err != nil {
-						return consumererror.NewLogs(err, *subLogs(&ld, bufFront))
-					}
-				}
-				buf.Reset()
-
-				// Writing truncated bytes back to buffer.
-				tmpBuf.WriteTo(buf)
-
-				bufFront, bufLen = nil, buf.Len()
+
+			var err error
+
+			if isProfilingData(ills.At(j)) {
+				err, profilingBufLen, profilingBufFront, permanentErrors = c.pushLogRecords(ctx, logs, res, i, j, profilingBuf, profilingEncoder, profilingTmpBuf, bufCap, profilingBufLen, profilingBufFront, permanentErrors, profilingHeaders, send)
```

Comment:
This function call is very hard to follow, 14 arguments and 4 return values is too much :)

- Please create a struct for buffering data to avoid passing i, j, profilingBuf, profilingEncoder, profilingTmpBuf, bufCap, profilingBufLen, profilingBufFront and returning profilingBufLen, profilingBufFront.
- Don't need to pass existing `permanentErrors`, we can return list of errors instead and merge it with `permanentErrors` after the call.

I suppose the function signature can be something like :
```
func (c *client) pushLogRecords(ctx context.Context, logs pdata.LogSlice, res pdata.Resource, buf *bufferState, headers map[string]string, send func(context.Context, *bytes.Buffer, map[string]string) error) (sendingError error, permanentErrors []error)
```

Reply:
Completely agreed. Did just that, thanks for the tips. So much cleaner now!

Reply:
I also re-ran the benchmarks, and the overhead is now 1-3 percentage points lower, that is around 1-2% for non-profiling data and 2-5% for profiling data in most cases. I guess stack pushes and pops do not come for free :)